app/interactors/shopping_cart.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+class ShoppingCart
+  include Dry::Monads[:result]
+
+  attr_reader :cart, :cart_identifier
+
+  def initialize(cart_identifier = nil)
+    @cart_identifier = cart_identifier
+    @cart = build_cart
+  end
+
+  def add_item(product_id:, quantity:)
+    if new_product?(product_id: product_id)
```

Comment:
this  if block is ugly - method has too many responsibilities. should be splitted